File: C/322-coin-change/main.go
package main

import (
	"fmt"
)

func coinChange(coins []int, amount int) int {
	buf := make([]int, amount+1)

	buf[amount] = 1
	for i := amount; i >= 1; i-- {
		if buf[i] == 0 {
			continue
		}
		for j := 0; j < len(coins); j++ {
			if i >= coins[j] && ((buf[i-coins[j]] > 0 && buf[i-coins[j]] > buf[i]+1) || buf[i-coins[j]] == 0) {
				buf[i-coins[j]] = buf[i] + 1
			}
		}
	}
	return buf[0] - 1
}

func main() {
	fmt.Println(coinChange([]int{1, 2, 5}, 11))
	fmt.Println(coinChange([]int{2}, 3))
}

File: C/322-coin-change/main.rs
struct Solution {}

impl Solution {
  pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
    let mut dp: Vec<i32> = vec![0; amount as usize + 1];
    dp[0] = 1;
    (0..amount as usize).for_each(|v| {
      if dp[v] == 0 {
        return;
      }
      coins.iter().for_each(|&c| {
        if c <= amount
          && c + v as i32 <= amount
          && (dp[c as usize + v] == 0 || dp[c as usize + v] > dp[v] + 1)
        {
          dp[c as usize + v] = dp[v] + 1;
        }
      });
    });

    dp[amount as usize] - 1
  }
}
